### PR TITLE
Amazon Q : Fixing the inappropriate yellow lines shown in the editor file using CodeSnippet.

### PR DESCRIPTION
**Pull request by laileni-aws**
## Problem
- If user triggers a scan and modifies the code in the editor, the security issue shows yellow squiggle lines at different place in the editor.

## Solution:
- To avoid this we compare the existing code diff from editor with the codeSnippet from response. If both matches we show the issue to the user if not we drop the suggestion.
- Implemented only for Auto-Scans.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/codewhisperer/service/securityScanHandler.ts [hunk 1]
```diff
@@ -109,19 +111,34 @@ function mapRawToCodeScanIssue(issue: RawCodeScanIssue): CodeScanIssue {
...
     const codeScanIssues: RawCodeScanIssue[] = JSON.parse(json)
     codeScanIssues.forEach(issue => {
-        if (codeScanIssueMap.has(issue.filePath)) {
-            const list = codeScanIssueMap.get(issue.filePath)
-            if (list === undefined) {
-                codeScanIssueMap.set(issue.filePath, [issue])
+        let shouldShow = true
+
+        if (scope === CodeWhispererConstants.CodeAnalysisScope.FILE && editor) {
+            shouldShow = issue.codeSnippet.every(code => {
```
Comment (ctlai95) [inline packages/core/src/codewhisperer/service/securityScanHandler.ts:125]: IIRC doesn't the code snippet include a few lines before and after? If so we need to check only the impacted line numbers

File: packages/core/src/codewhisperer/service/securityScanHandler.ts [hunk 2]
```diff
@@ -109,19 +111,34 @@ function mapRawToCodeScanIssue(issue: RawCodeScanIssue): CodeScanIssue {
...
+        }
+
+        if (shouldShow) {
+            const filePath = issue.filePath
+            if (codeScanIssueMap.has(filePath)) {
+                const issueList = codeScanIssueMap.get(filePath)
+                if (issueList) {
+                    issueList.push(issue)
+                    codeScanIssueMap.set(filePath, issueList)
+                }
```
Comment (ctlai95) [inline packages/core/src/codewhisperer/service/securityScanHandler.ts:138]: There's an else block missing, is that intentional?

```typescript
codeScanIssueMap.set(filePath, [issue])
```

File: packages/core/src/codewhisperer/service/securityScanHandler.ts [hunk 3]
```diff
@@ -109,19 +111,34 @@ function mapRawToCodeScanIssue(issue: RawCodeScanIssue): CodeScanIssue {
     }
 }
 
-function mapToAggregatedList(codeScanIssueMap: Map<string, RawCodeScanIssue[]>, json: string) {
+function mapToAggregatedList(
```
Comment (ctlai95) [inline packages/core/src/codewhisperer/service/securityScanHandler.ts:114]: please add some unit tests for the new logic

File: packages/core/src/codewhisperer/service/securityScanHandler.ts (hunk 1 above)
Comment (laileni-aws) [reply packages/core/src/codewhisperer/service/securityScanHandler.ts:125]: Right now,
I am checking all the text provided in the code snippet. Should I restrict this to only the impacted line numbers ?

File: packages/core/src/codewhisperer/service/securityScanHandler.ts (hunk 2 above)
Comment (laileni-aws) [reply packages/core/src/codewhisperer/service/securityScanHandler.ts:138]: I think Both code snippet does the same functionality

File: packages/core/src/codewhisperer/models/errors.ts [hunk 4]
```diff
@@ -30,8 +30,8 @@ export class ProjectSizeExceededError extends SecurityScanError {
 }
 
 export class DefaultError extends SecurityScanError {
-    constructor() {
-        super('Security scan failed.', 'DefaultError', DefaultCodeScanErrorMessage)
+    constructor(error: string | undefined) {
+        super(`Security scan failed. ${error}`, 'DefaultError', DefaultCodeScanErrorMessage)
```
Comment (justinmk3) [inline packages/core/src/codewhisperer/models/errors.ts:34]: why not pass the error as the `cause` field? also consider `ToolkitError.chain()` .

strinigifying errors randomly in messages completely defeats the builtin mechanisms in the extension such as : https://github.com/aws/aws-toolkit-vscode/blob/6300d520ac347b7596d98c7cec312b67f98ee528/packages/core/src/shared/errors.ts#L467

and makes it much harder to show useful errors to users, in reporting, etc. Please, use care when designing errors.

Comment (laileni-aws) [reply packages/core/src/codewhisperer/models/errors.ts:34]: These errors are not shown to the user. This is for telemetry purpose only.

Comment (laileni-aws) [reply packages/core/src/codewhisperer/models/errors.ts]: This is redundant change. Reverted acc in draft CR

File: packages/core/src/codewhisperer/service/securityScanHandler.ts [hunk 5]
```diff
@@ -109,19 +111,32 @@ function mapRawToCodeScanIssue(issue: RawCodeScanIssue): CodeScanIssue {
...
+                }
             }
+        }
+        return true
+    })
+
+    filteredIssues.forEach(issue => {
+        const filePath = issue.filePath
+        if (codeScanIssueMap.has(filePath)) {
+            codeScanIssueMap.get(filePath)!.push(issue)
```
Comment (ctlai95) [inline packages/core/src/codewhisperer/service/securityScanHandler.ts:137]: Can we avoid using `!`? I think it is not guaranteed that the map entry contains an array type

**Review by ctlai95 — APPROVED** [review]

File: packages/core/src/codewhisperer/models/errors.ts (hunk 4 above)
Comment (justinmk3) [reply packages/core/src/codewhisperer/models/errors.ts]: > These errors are not shown to the user

This cannot be assumed. Codepaths can change. Errors should use best practices.

**Comment by justinmk3** [comment]
Does this have a customer-facing impact that can be described in the changelog?

1. Review the [changelog guidelines](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#changelog).
2. Add a new changelog entry: `npm run newChange -w packages/amazonq` or `npm run newChange -w packages/toolkit`

**Pull request merged**